### ISSUE TITLE
Integrating Reticulum Network Stack in the Future

### ISSUE THREAD

**Issue by Sudo-Ivan** [s2underground/GhostNet#26]
# What is the Reticulum Network Stack?

The Reticulum Network Stack is a decentralized cryptography-based network stack that allows communication over any medium, including HF/VHF/UHF, LoRa, Internet, and virtually anything if you create the software support for it. It supports low bandwidth, high latency networks, which is perfect for something like HF or LoRa.

> The vision of Reticulum is to allow anyone to be their own network operator, and to make it cheap and easy to cover vast areas with a myriad of independent, inter-connectable and autonomous networks. Reticulum is not one network. It is a tool for building thousands of networks. Networks without kill-switches, surveillance, censorship and control. Networks that can freely interoperate, associate and disassociate with each other, and require no central oversight. Networks for human beings. Networks for the people. - https://github.com/markqvist/Reticulum

There is still more work that needs to be done on the stack, and the cryptography needs to be reviewed by a professional. The purpose of this issue is to discuss how to implement it in the future.

# How will it integrate with the GhostNet?

![rns-ghostnet](https://github.com/user-attachments/assets/b58f873a-a1bd-4011-b4cc-2d394a0d0e06)

The stack can combine different mediums to achieve a wide range of coverage and complicate censorship efforts greatly.

## Internet

The stack supports TCP/UDP (regular internet) and [I2P](https://geti2p.net/en/) (Invisible Internet Project), so a community network can be created to ensure censorship-resistant comms and exchange anything from files, memes, and voice messages. A TCP/I2P node would need to be created to allow users to find each other. The data is not decrypted at the server; if propagation is enabled, it can store encrypted messages for offline users for up to 30 days but the data remains encrypted until reaching the user.

A cheap VPS can be used to create the node for the network or self-hosted over I2P using something as cheap as a Raspberry Pi Zero W 2. However, users would need to install and setup I2P to be able to use that interface. Additional existing networks can be used as a backup if the node goes down. Even [private networks](https://markqvist.github.io/Reticulum/manual/understanding.html#interface-access-codes) can be created, and only those with the passphrase can join.

## LoRa

[Reticulum RNode](https://markqvist.github.io/Reticulum/manual/hardware.html#rnode) provides LoRa support but requires a computer (such as a Raspberry Pi Zero W 2) as a full standalone version of the software in C++ is still a work in progress. This will allow groups to create private, secure groups and also combine with other interfaces like HF for long-range comms between RNode devices.

## HF/VHF/UHF

In the event of a dire national emergency, and if necessary, secure comms can be established over radio to exchange sensitive information.

Someone tested it over HF, here is a video: https://youtu.be/blwNVumLujc?si=vID7p5BuPvoIzIK7

## Support for Other Communication Methods

A custom interface can be created to support really any communication method that at minimum supports half-duplex operation, provides an average throughput of 5 bits per second or greater, and supports a physical layer MTU of 500 bytes.

# Communication possibilities

## Chat

You can send files, images and voice messages.

## Pages (Very minimal websites)

Any user can host a node and create a page using markup. Like this:

![image](https://github.com/user-attachments/assets/472393de-066f-4173-94e8-eadd18fd346d)

Even dynamic node generated pages using PHP, Python and bash.

## Email

There is no email client yet, but it is very possible and potentially easy to build one that I might take on the task of creating it within the next few months if no one else does.

## Voice 

MeshChat and Sideband support exchanging voice messages and MeshChat has a calling feature so you can call other users on MeshChat, even over LoRa.

![image](https://github.com/user-attachments/assets/239467bc-d766-485e-b4b6-2bd58464ff11)

## ATAK and FreeTAK

Community members, including myself have talked about creating an ATAK integration.  

The FreeTAK team as adopted Reticulum stack into their flock of parrots concept: https://matrix.to/#/!TRaVWNnQhAbvuiSnEK:matrix.org/$NSp9xNVnbFqG7ezLQGNTkmtVLyVV8bNssT65tq8q5lw?via=matrix.org&via=tchncs.de&via=matrix.minuteman-tech.net

# Possible Issues Blocking Integration (as of 02 OCT 2024)

These are potential issues that may prevent adoption:

1. **Group Messaging**
2. **Barrier to entry**
3. **Apps and Integrations**
4. **Hardware Support**
5. **Simple Guides and Video Tutorials**

# For users that want to test it now
Current nodes and users
![image](https://github.com/user-attachments/assets/5029deed-6d88-4eaf-99b7-f1585ec6c2d7)

## Existing Apps

[Meshchat](https://github.com/liamcottle/reticulum-meshchat) - Web/Electron app 
[NomadNet](https://github.com/markqvist/nomadnet) - Terminal Interface
[Sideband](https://github.com/markqvist/NomadNet) - Android and [IOS Testflight](https://testflight.apple.com/join/ftgW9n0Z)

## Awesome Reticulum

https://github.com/markqvist/Reticulum/wiki/Awesome-Reticulum

## Manual, Videos and FAQ

I barely touched the surface of the stack and really suggest anyone who is interested read the full manual here: https://markqvist.github.io/Reticulum/manual/

If you want a short video explanation of the Reticulum Network Stack, I recommend checking out: https://www.youtube.com/watch?v=q8ltLt5SK6A

FAQ can be found here: https://github.com/markqvist/Reticulum/wiki/Frequently-Asked-Questions

## Matrix Group for Reticulum

Join the official matrix room here: https://matrix.to/#/#reticulum:matrix.org

# How am I contributing?

I am working on [several bots](https://gitlab.com/ivans-reticulum-bots), hardware encryption support (for Kryptor FPGA), donating, and also have some apps planned for the future. Occasionally I will do some security testing and make sure other projects are implementing best coding practices to prevent exploits.

**Comment by tim292stro** [comment]
I support this effort, and agree with the conclusions made above. I moved from Meshtestic to rNode a while back and have done some of my own ATAK server work (a "superuser" node on an RPi Zero 2 w as a USB-C connected local "server", providing longer reach connectivity to normal users in range).  I'm interested if this gets traction in GhostNet, and would be willing to help out.

**Comment by Sudo-Ivan** [comment]
Thanks for reminding me, The offline features are very useful in certain situations.

**Comment by s2underground** [comment]
Yes, reticulum will be integrated into this in the future. For now, a lot of content has to be made on it that explains how things work for the layperson, but it will be integrated at some point

**Comment by Sudo-Ivan** [comment]
# Update

## Videos

[Reticulum Network Stack Youtube Playlist](https://youtube.com/playlist?list=PLLJCVvgWkWBB8NxT3NNTSSeaHisXRZXtr&si=kUbHEhowkzTWpsDD)

[Developer Mark Talks at 38C3 2024](https://media.ccc.de/v/38c3-reticulum-unstoppable-networks-for-the-people)

**Comment by Sudo-Ivan** [comment]
[LXST framework](https://github.com/markqvist/lxst) - Lightweight Extensible Signal Transport for Reticulum.

https://hub.federated.channel/channel/reticulum?mid=5f06882f-ddbe-4cf0-ab9c-bac79f9172ff

![Reticulum Phone](https://hub.federated.channel/photo/5e7d5877-bef9-4b21-89a1-9de9ef23eb68-2.jpg)

**Comment by Sudo-Ivan** [comment]
https://unsigned.io/articles/2025_05_09_The_End_Is_Nigh_For_The_Beta_Days.html

V1.0.0 is near. As of Reticulum version 0.9.5:

- Massive performance and memory optimizations
- Much faster I/O backend
- Significantly expanded hardware support
- Complete interface modularity
- Migrating to AES-256 encryption for all communication
- A new real-time voice protocol for Reticulum [LXST](https://github.com/markqvist/lxst)
- Much higher capacity for LXMF propagation networks
- Significant improvements to overall usability and utilities

**Comment by Sudo-Ivan** [comment]
Late Update:

Reticulum v1.0.0 was released on 14 July. The developer will be taking a break for a bit, as mentioned in the blog post https://unsigned.io/articles/2025_05_09_The_End_Is_Nigh_For_The_Beta_Days.html.

**Comment by Sudo-Ivan** [comment]
My current projects:

[Reticulum-Go](https://github.com/Sudo-Ivan/Reticulum-Go) - Go implementation of Reticulum Network Stack.
[LXMFy](https://github.com/lxmfy/LXMFy) - LXMF Bot framework.
[rns-page-node](https://github.com/Sudo-Ivan/rns-page-node) - Simple way to serve pages/files .
[Ren Browser](https://github.com/Sudo-Ivan/Ren-Browser) - Mobile/Desktop browser for Reticulum pages. 
[RNS-over-HTTP](https://github.com/Sudo-Ivan/RNS-over-HTTP) - HTTP Interface for Reticulum.

Working on guides in english and russian as I build and learn more things about the network stack.

**Comment by Sudo-Ivan** [comment]
Reticulum Basics for Everyday Users - by Emcomm Solutions:

https://www.youtube.com/watch?v=UkGo1BZSuaE

**Comment by Sudo-Ivan** [comment]
Sideband LXST Voice Calls For Android - https://github.com/markqvist/Reticulum/discussions/978
